proposals/large-ipld-dags.md

```diff
@@ -0,0 +1,116 @@
+# Support Large IPLD/IPFS DAGs
+
+Authors: Stebalien
+
+Initial PR: TBD <!-- Reference the PR first proposing this document. Oooh, self-reference! -->
+
+## Purpose &amp; impact 
+#### Background &amp; intent
+_Describe the desired state of the world after this project? Why does that matter?_
+
+First, it should be possible to store arbitrary and arbitrarily large IPLD DAGs on Filecoin using
+the built-in protocols. At the moment, Filecoin can only store "whole DAGs". If a DAG, doesn't fit
+into a sector when serialized as a CAR, it must be converted to raw-blocks, chunked, and then stored
+as those chunks.
+
+Unfortunately:
+
+1. This workaround erases the underlying DAG structure. This makes it difficult to transfer this
+   data for both storage and retrieval. This is especially true when interacting with IPFS.
+2. This workaround requires storing an "overlay" DAG in Filecoin (paying for that storage).
+
+Second, it should be possible to retrieve subsets of DAGs. While the underlying protocols support
+retrieving subsets of DAGs, the CLI does not. This makes it impossible to, e.g., retrieve a single
+file from a directory without modifying Lotus.
+
+#### Assumptions &amp; hypotheses
+_What must be true for this project to matter?_
+
+There is no easy way (e.g., no out-of-band deals) to store large (> sector size) IPLD DAGs while
+preserving the DAG structure.
+
+#### User workflow example
+_How would a developer or user use this new capability?_
+
+* `lotus client deal` should accept an IPLD selector.
+* `lotus cleint deal` should automatically split large DAGs between multiple sectors.
+* `lotus client retrieve` should support retrieving IPLD selectors (dag subsets).
+
+#### Impact
+_How directly important is the outcome to web3 dev stack product-market fit?_
+
+🔥
+
+At the moment, any tool wishing to support storing IPFS files/directories larger than 32GiB will need to store these IPFS files/directories as "raw blocks", throwing away all the DAG structural information. This will make future retrieval deals for subsets of this data infeasible and will make IPFS interop extremely difficult.
+
+This is only one 🔥 because there are plenty of useful sub-32GiB datasets and non-IPFS datasets.
```

Comment:
This is true although there is additional impact here which is enabling people to store compositions of data sets.

If deals already exist on Filecoin for a dataset and then someone wants to reference that dataset (or some part of it) within theirs then the data has to be duplicated and stored in two separate deals. With this feature as long as there is a way to discover mappings of CID -> miner with CID (currently out of band, but is a required part of retrieval market work) then users don't need to store the same data twice (or worry about compositions exceeding 32GiB)